src/core/dom/layout/viewport-observer.js

```diff
@@ -70,22 +83,32 @@ export function observeIntersections(element, viewportCallback) {
     callbacks = [];
     viewportCallbacks.set(element, callbacks);
   }
-
-  callbacks.push(viewportCallback);
+  callbacks.push(callback);
   viewportObserver.observe(element);
 }
 
 /**
- * Unobserve an element.
+ * Unobserves the intersection observer for the given callback.
+ * If no callbacks remain for the element, unobserves the element too.
+ *
  * @param {!Element} element
+ * @param {function(IntersectionObserverEntry)} callback
  */
-export function unobserveWithSharedInOb(element) {
-  const win = toWin(element.ownerDocument.defaultView);
-  const viewportObserver = viewportObservers.get(win);
-  viewportObserver?.unobserve(element);
-  // TODO(dmanek): This is a potential bug. We only want to remove
-  // a single callback as opposed to all.
-  viewportCallbacks.delete(element);
+export function unobserveIntersections(element, callback) {
+  const callbacks = viewportCallbacks.get(element);
+  if (!callbacks) {
+    return;
+  }
+  const idxToDelete = callbacks.findIndex((cb) => cb === callback);
```

Comment:
Can we rely on checking function equality this way? Does this require elements using the shared intersection observer to store a reference to their callback?

Reply:
Yes we can. Functions in JS are objects and stored by reference, so we can compare using `===`. For example, this works:
```
function foo() {};
let a = new Map();
a.set('bar', foo);
let b = new Map();
b.set('baz', foo);
a.get('bar') === b.get('baz') // true
```

Reply:
>  Does this require elements using the shared intersection observer to store a reference to their callback?

Yes

Reply:
We do that here: https://github.com/ampproject/amphtml/blob/main/src/core/dom/layout/viewport-observer.js#L33

Reply:
Thanks for the clarifications. I mostly mean that there are some uses of `observeWithSharedInOb` that pass in an anonymous function, and we should be careful that they pass in a fn reference when unobserving because we can't deduce equality otherwise:
```javascript
(() => {}) === (() => {}) // false
```

Just a note for the migration step.

Reply:
The other common alternative is to return a cleanup closure:

```js
const cleanup = observeIntersections(…);

// later
cleanup();
```

This is really simple in `observeIntersections`:

```js
function observeIntersections(element, callback) {
  // …

  return () => {
    unobserveIntersections(element, callback);
  };
}

Reply:
Good point. Thanks!

Reply:
@caroqliu thanks - will keep that in mind.

@jridgewell that's a great idea! However, due to legacy usage of `observeWithSharedInOb` (which wraps `observeIntersections`) this will be a breaking change. I've updated #36088 to include this.